### DISCUSSION TITLE
LOLIN S2 Mini - SD Card support ?

### DISCUSSION THREAD

**Discussion by wvthoog** [micropython/micropython#10469] [ESP32]
Hi,

so i'm trying to get the Lolin SD card shield to work with a Lolin S2 Mini and decided to rebuild the firmware to include it (since it's not enabled by default) I'm using ESP-IDF version 4.4 and the latest clone of Micropython 

Then i edited the file mpconfigboard.h in the ports/esp32/boards/LOLIN_S2_MINI directory to enable the sdcard 

`#define MICROPY_HW_ENABLE_SDCARD            (1)`

and started the build

`idf.py -D MICROPY_BOARD=LOLIN_S2_MINI build`

Unfortunately it fails with the following error

`/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c: In function 'machine_sdcard_make_new':
/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c:185:35: error: implicit declaration of function 'SDMMC_HOST_DEFAULT'; did you mean 'SDSPI_HOST_DEFAULT'? [-Werror=implicit-function-declaration]
         sdmmc_host_t _temp_host = SDMMC_HOST_DEFAULT();
                                   ^~~~~~~~~~~~~~~~~~
                                   SDSPI_HOST_DEFAULT
/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c:185:35: error: invalid initializer
/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c:194:50: error: 'VSPI_HOST' undeclared (first use in this function); did you mean 'FSPI_HOST'?
         self->host.slot = slot_num ? HSPI_HOST : VSPI_HOST;
                                                  ^~~~~~~~~
                                                  FSPI_HOST
/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c:194:50: note: each undeclared identifier is reported only once for each function it appears in
/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c:222:30: error: 'GPIO_NUM_23' undeclared (first use in this function); did you mean 'GPIO_NUM_43'?
                 .gpio_mosi = GPIO_NUM_23,
                              ^~~~~~~~~~~
                              GPIO_NUM_43
/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c:248:9: error: unknown type name 'sdmmc_slot_config_t'; did you mean 'sdspi_slot_config_t'?
         sdmmc_slot_config_t slot_config = SDMMC_SLOT_CONFIG_DEFAULT();
         ^~~~~~~~~~~~~~~~~~~
         sdspi_slot_config_t
/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c:248:43: error: implicit declaration of function 'SDMMC_SLOT_CONFIG_DEFAULT'; did you mean 'SDSPI_SLOT_CONFIG_DEFAULT'? [-Werror=implicit-function-declaration]
         sdmmc_slot_config_t slot_config = SDMMC_SLOT_CONFIG_DEFAULT();
                                           ^~~~~~~~~~~~~~~~~~~~~~~~~
                                           SDSPI_SLOT_CONFIG_DEFAULT
/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c:83:11: error: request for member 'gpio_cd' in something not a structure or union
     config.pin_var = pin_or_int(arg_vals[arg_id].u_obj)
           ^
/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c:254:9: note: in expansion of macro 'SET_CONFIG_PIN'
         SET_CONFIG_PIN(slot_config, gpio_cd, ARG_cd);
         ^~~~~~~~~~~~~~
/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c:83:11: error: request for member 'gpio_wp' in something not a structure or union
     config.pin_var = pin_or_int(arg_vals[arg_id].u_obj)
           ^
/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c:255:9: note: in expansion of macro 'SET_CONFIG_PIN'
         SET_CONFIG_PIN(slot_config, gpio_wp, ARG_wp);
         ^~~~~~~~~~~~~~
/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c:259:24: error: request for member 'width' in something not a structure or union
             slot_config.width = width;
                        ^
/home/wvthoog/projects/other/audiotour/micropython/ports/esp32/machine_sdcard.c:265:23: error: implicit declaration of function 'sdmmc_host_init_slot'; did you mean 'sdspi_host_init_slot'? [-Werror=implicit-function-declaration]
         check_esp_err(sdmmc_host_init_slot(self->host.slot, &slot_config));
                       ^~~~~~~~~~~~~~~~~~~~
                       sdspi_host_init_slot
cc1: some warnings being treated as errors
[1296/1302] Building C object esp-idf/main/CMakeFiles/__idf_main.dir/home/wvthoog/projects/other/audiotour/micropython/lib/littlefs/lfs2.c.obj
ninja: build stopped: subcommand failed.
ninja failed with exit code 1`

So my question is, does Micropython actually support sd cards for the Lolin S2 Mini ? And secondly, do i need to define more variables in mpconfigboard.h ? If so, which ones 

Thanks for reading

Regards,
Wim

**Comment by jimmo** [comment]
@wvthoog  The IDF sdcard driver is not currently supported on S2.

If you're interested in making this work, the reason is that the S2 doesn't support MMC, and our driver doesn't check that. (It needs to use the `SOC_SDMMC_HOST_SUPPORTED` to disable the MMC code).

It also possibly needs to change some other defaults, see a similar change for S3 support:
https://github.com/micropython/micropython/commit/1904833e0c7c650e86d8e151ea57ea1fa2ab09bd

**Reply by robert-hh** [reply]
Until this is added to the firmware, you should be able to use the Python driver sdcard.py at https://github.com/micropython/micropython-lib/tree/master/micropython/drivers/storage/sdcard. It uses SPI for the data transfer and works well.

**Reply by wvthoog** [reply]
@jimmo That sounds like the route i want to go. If that fails, i'll go with @robert-hh solution. Which already worked, but i rather want a custom firmware with sdcard support built in. 

So, from what I've read the file machine_sdcard.c in ports/esp32 supports both MMC and SPI mode. Of which only SPI is supported by the ESP32S2. As you suggested I've disabled MMC support in the ESP-IDF by editing the file soc_caps.h in esp-idf/components/soc/esp32s2/include/soc/ and added:

`SOC_SDMMC_HOST_SUPPORTED (0)`

Rebuilt the framework and started the Micropython built again, but failed the same way as before. 

Could you elaborate a bit more how to disable the MMC part of the code but maintain SPI support.

**Reply by jimmo** [reply]
Yes, sorry I meant to mention sdcard.py and then went digging through the IDF code and forgot about it. Thanks @robert-hh 

> As you suggested I've disabled MMC support in the ESP-IDF

Not quite what I meant sorry. The IDF will automatically disable/enable MMC support based on the SoC caps. What you need to do is make the MicroPython SDCard driver aware of the SoC caps (using the macros set).

**Reply by wvthoog** [reply]
Yeah, that would be way above my knowledge skills. Will just wait for it to be implemented in future releases. Until then i'll use the sdcard.py solution. Thanks for replying
